File: test_1d4_crc_reference.py
#!/usr/bin/env python3
"""
0x1D4 CRC Validation Test
Tests our CRC algorithm against known-good reference frames from Thunderstruck TVCU
"""

def crc8_msb_poly_1d29(data):
    """CRC-8 MSB with polynomial 0x1D29 for 0x1D4 frames"""
    crc = 0
    for byte in data:
        crc ^= byte
        for _ in range(8):
            if crc & 0x80:
                crc = ((crc << 1) ^ 0x1D) & 0xFF
            else:
                crc = (crc << 1) & 0xFF
    return crc

def compute_exact_1d4_wire_crc(payload8):
    """Wire CRC for 0x1D4 frame (ID + payload)"""
    frame = bytes([0xD4]) + payload8[:7]
    return crc8_msb_poly_1d29(frame)

def main():
    print("=== 0x1D4 CRC Reference Frame Validation ===")
    print("Known-good frames from Thunderstruck TVCU\n")
    
    # Reference frames with known-good CRC
    ref_frames = [
        {
            'data': bytes.fromhex("6E 6E 00 00 87 44 01 23"),
            'expected_crc': 0x23,
            'description': "Counter=8 (high), Charge=0x44, Byte5=0x87"
        },
        {
            'data': bytes.fromhex("6E 6E 00 00 C7 44 01 E4"),
            'expected_crc': 0xE4,
            'description': "Counter=C (high), Charge=0x44, Byte5=0xC7"
        },
        {
            'data': bytes.fromhex("6E 6E 00 00 07 44 01 28"),
            'expected_crc': 0x28,
            'description': "Counter=0 (high), Charge=0x44, Byte5=0x07"
        },
        {
            'data': bytes.fromhex("6E 6E 00 00 47 44 01 EF"),
            'expected_crc': 0xEF,
            'description': "Counter=4 (high), Charge=0x44, Byte5=0x47"
        },
    ]
    
    pass_count = 0
    fail_count = 0
    
    for idx, frame in enumerate(ref_frames):
        frame_data = frame['data']
        expected_crc = frame['expected_crc']
        
        # Calculate CRC using our algorithm
        calculated_crc = compute_exact_1d4_wire_crc(frame_data)
        
        # Display frame
        print(f"[Frame {idx+1}] {frame['description']}")
        print(f"  Data: {' '.join(f'{b:02X}' for b in frame_data)}")
        print(f"  Expected CRC:   0x{expected_crc:02X}")
        print(f"  Calculated CRC: 0x{calculated_crc:02X}")
        
        if calculated_crc == expected_crc:
            print("  ✓ PASS\n")
            pass_count += 1
        else:
            print("  ✗ FAIL - Mismatch!\n")
            fail_count += 1
    
    print("=== Summary ===")
    print(f"Passed: {pass_count}/{len(ref_frames)}")
    print(f"Failed: {fail_count}/{len(ref_frames)}")
    
    if fail_count == 0:
        print("\n✓ All tests PASSED! CRC algorithm is CORRECT.")
        print("\nKey findings:")
        print("  • Byte 5 low nibble = 0x7 (not 0x4!)")
        print("  • Byte 6 (charge status) = 0x44 (not 0x1E or 0x30!)")
        print("  • Counter in byte 5 high nibble cycles: 8→C→0→4")
        return 0
    else:
        print("\n✗ Tests FAILED! CRC algorithm needs adjustment.")
        return 1

if __name__ == "__main__":
    exit(main())

File: test_1d4_crc_reference.cpp
#include <cstdio>
#include <cstdint>
#include <cstring>

// CRC-8 MSB with polynomial 0x1D29 (for 0x1D4 frames)
// This is the algorithm we're using in the firmware
uint8_t crc8MsbPoly1D29(const uint8_t* data, uint32_t len) {
    uint8_t crc = 0U;
    for (uint32_t i = 0U; i < len; i++) {
        crc ^= data[i];
        for (uint32_t j = 0U; j < 8U; j++) {
            if (crc & 0x80U) {
                crc = (uint8_t)((crc << 1U) ^ 0x1DU);  // 0x1D29 >> 8 = 0x1D
            } else {
                crc = (uint8_t)(crc << 1U);
            }
        }
    }
    return crc;
}

// Wire CRC algorithm for 0x1D4 (same as 0x1DA but with different poly)
uint8_t computeExact1d4WireCrc(uint8_t idLo, const uint8_t* payload8) {
    uint8_t p[8] = {0U};
    p[0] = idLo;
    memcpy(&p[1], payload8, 7U);
    return crc8MsbPoly1D29(p, 8U);
}

int main(void) {
    printf("=== 0x1D4 CRC Reference Frame Validation ===\n");
    printf("Known-good frames from Thunderstruck TVCU\n\n");
    
    // Reference frames with known-good CRC
    const struct {
        uint8_t data[8];
        uint8_t expectedCrc;
        const char* description;
    } refFrames[] = {
        {{0x6E, 0x6E, 0x00, 0x00, 0x87, 0x44, 0x01, 0x23}, 0x23, "Counter=8 (high), Charge=0x44, Byte5=0x87"},
        {{0x6E, 0x6E, 0x00, 0x00, 0xC7, 0x44, 0x01, 0xE4}, 0xE4, "Counter=C (high), Charge=0x44, Byte5=0xC7"},
        {{0x6E, 0x6E, 0x00, 0x00, 0x07, 0x44, 0x01, 0x28}, 0x28, "Counter=0 (high), Charge=0x44, Byte5=0x07"},
        {{0x6E, 0x6E, 0x00, 0x00, 0x47, 0x44, 0x01, 0xEF}, 0xEF, "Counter=4 (high), Charge=0x44, Byte5=0x47"},
    };
    
    const uint32_t numFrames = sizeof(refFrames) / sizeof(refFrames[0]);
    uint32_t passCount = 0;
    uint32_t failCount = 0;
    
    for (uint32_t i = 0; i < numFrames; i++) {
        const uint8_t* frameData = refFrames[i].data;
        uint8_t expectedCrc = refFrames[i].expectedCrc;
        
        // Calculate CRC using our algorithm
        uint8_t calculatedCrc = computeExact1d4WireCrc(0xD4U, frameData);
        
        // Display frame
        printf("[Frame %u] %s\n", i+1, refFrames[i].description);
        printf("  Data: %02X %02X %02X %02X %02X %02X %02X %02X\n",
               frameData[0], frameData[1], frameData[2], frameData[3],
               frameData[4], frameData[5], frameData[6], frameData[7]);
        printf("  Expected CRC:   0x%02X\n", expectedCrc);
        printf("  Calculated CRC: 0x%02X\n", calculatedCrc);
        
        if (calculatedCrc == expectedCrc) {
            printf("  ✓ PASS\n\n");
            passCount++;
        } else {
            printf("  ✗ FAIL - Mismatch!\n\n");
            failCount++;
        }
    }
    
    printf("=== Summary ===\n");
    printf("Passed: %u/%u\n", passCount, numFrames);
    printf("Failed: %u/%u\n", failCount, numFrames);
    
    if (failCount == 0) {
        printf("\n✓ All tests PASSED! CRC algorithm is CORRECT.\n");
        return 0;
    } else {
        printf("\n✗ Tests FAILED! CRC algorithm needs adjustment.\n");
        return 1;
    }
}
